pkg/instance/backupService.go

```diff
@@ -0,0 +1,240 @@
+	CompleteMultipartUpload(ctx context.Context, params *s3.CompleteMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.CompleteMultipartUploadOutput, error)
+	AbortMultipartUpload(ctx context.Context, params *s3.AbortMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.AbortMultipartUploadOutput, error)
+}
+
+// BackupStats tracks backup operation statistics
+type BackupStats struct {
+	ObjectsProcessed int64
+	BytesProcessed   int64
+	StartTime        time.Time
+	mu               sync.Mutex
+}
+
+func (bs *BackupStats) increment(size int64) {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+	bs.ObjectsProcessed++
+	bs.BytesProcessed += size
+}
+
+// PerformBackup executes the backup operation
+func (s *BackupService) PerformBackup(ctx context.Context, minioBucket, s3Bucket, key string) error {
+	g, ctx := errgroup.WithContext(ctx)
+	stats := &BackupStats{StartTime: time.Now()}
+	pr, pw := io.Pipe()
+
+	g.Go(func() error {
+		defer pw.Close()
+		return s.createTarGzStream(ctx, pw, minioBucket, stats)
+	})
+
+	g.Go(func() error {
+		return s.streamToS3WithMultipart(ctx, s3Bucket, key, pr)
+	})
+
+	if err := g.Wait(); err != nil {
+		return fmt.Errorf("backup failed: %v", err)
+	}
+
+	s.logBackupStats(stats)
+	return nil
+}
+func (s *BackupService) createTarGzStream(ctx context.Context, w io.Writer, minioBucket string, stats *BackupStats) error {
+	gw := gzip.NewWriter(w)
+	defer gw.Close()
+
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	objectCh := s.minioClient.ListObjects(ctx, minioBucket, minio.ListObjectsOptions{
+		Recursive: true,
+	})
+
+	for object := range objectCh {
+		if object.Err != nil {
+			return fmt.Errorf("list objects: %v", object.Err)
+		}
+
+		if err := s.processSingleObject(ctx, tw, minioBucket, object, stats); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *BackupService) processSingleObject(ctx context.Context, tw *tar.Writer, minioBucket string, object minio.ObjectInfo, stats *BackupStats) error {
+	s.logger.InfoContext(ctx, "Processing object", "key", object.Key, "bucket", minioBucket)
+	obj, err := s.minioClient.GetObject(ctx, minioBucket, object.Key, minio.GetObjectOptions{})
+	if err != nil {
+		return fmt.Errorf("get object %s: %v", object.Key, err)
+	}
+	defer obj.Close()
+
+	header := &tar.Header{
+		Name:    object.Key,
+		Size:    object.Size,
+		Mode:    0644,
+		ModTime: object.LastModified,
+	}
+
+	if err := tw.WriteHeader(header); err != nil {
+		return fmt.Errorf("write tar header for %s: %v", object.Key, err)
+	}
+
+	if _, err := io.Copy(tw, obj); err != nil {
+		return fmt.Errorf("copy object %s to tar: %v", object.Key, err)
+	}
+
+	stats.increment(object.Size)
+	return nil
+}
+
+func (s *BackupService) streamToS3WithMultipart(ctx context.Context, bucket, key string, reader io.Reader) error {
+	createResp, err := s.s3Client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
```

Comment:
I feel like shortening "Response" is pointless here.
```suggestion
	createResponse, err := s.s3Client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
```